flink-connectors/flink-connector-datagen/src/test/java/org/apache/flink/connector/datagen/source/DataGeneratorSourceITCase.java

```diff
@@ -195,40 +195,31 @@ void testGatedRateLimiter() throws Exception {
 
         final DataStreamSource<Long> streamSource =
                 env.fromSource(generatorSource, WatermarkStrategy.noWatermarks(), "Data Generator");
-        final DataStream<Tuple2<Integer, Long>> map =
-                streamSource.map(new SubtaskAndCheckpointMapper());
-        final List<Tuple2<Integer, Long>> results = map.executeAndCollect(1000);
-
-        final Map<Tuple2<Integer, Long>, Integer> collect =
-                results.stream()
-                        .collect(
-                                Collectors.groupingBy(
-                                        x -> (new Tuple2<>(x.f0, x.f1)), summingInt(x -> 1)));
-        for (Map.Entry<Tuple2<Integer, Long>, Integer> entry : collect.entrySet()) {
-            assertThat(entry.getValue()).isEqualTo(capacityPerSubtaskPerCycle);
-        }
+        final DataStream<Long> map = streamSource.flatMap(new FirstCheckpointFilter());
+        final List<Long> results = map.executeAndCollect(1000);
+
+        assertThat(results).hasSize(capacityPerCycle);
     }
 
-    private static class SubtaskAndCheckpointMapper
-            extends RichMapFunction<Long, Tuple2<Integer, Long>> implements CheckpointListener {
+    private static class FirstCheckpointFilter
+            implements FlatMapFunction<Long, Long>, CheckpointedFunction {
 
-        private long checkpointId = 0;
-        private int subtaskIndex;
+        private volatile boolean firstCheckpoint = true;
 
         @Override
-        public void open(Configuration parameters) {
-            subtaskIndex = getRuntimeContext().getIndexOfThisSubtask();
+        public void flatMap(Long value, Collector<Long> out) throws Exception {
+            if (firstCheckpoint) {
+                out.collect(value);
+            }
         }
 
         @Override
-        public Tuple2<Integer, Long> map(Long value) {
-            return new Tuple2<>(subtaskIndex, checkpointId);
+        public void snapshotState(FunctionSnapshotContext context) throws Exception {
```

Comment:
> Additionally, the RateLimitedSourceReader may reset the checkpoint limit at the wrong time. We don't really that to happen when the checkpoint is complete, but rather when the next checkpoint starts (== when snapshotState was called).

I don't understand that reasoning: Why do we stop processing the record when a new checkpoint is triggered rather than when the actual checkpoint is committed? :thinking: Could you elaborate here a bit more?

Reply:
What we want to happen is that at most X records are emitted per checkpoint.
`emit|emit|snapshot|emit|emit|snapshot`

In the ideal world, this happens:
`emit|emit|snapshot|notify|emit|emit|...`

But since `notifyCheckpointComplete` can arrive at arbitrary times, or not at all, there are some edge-cases.
We may not emit anything for one checkpoint if the rpc is lost:
`emit|emit|snapshot|snapshot|notify|emit|emit|...`
We may emit more records per checkpoint if the rpc is late:
`emit|emit|snapshot|snapshot|notify|emit|notify|emit|emit|...`

Reply:
ok, now I know what confused me. I was worried about the individual records not being included in the most-recent checkpoint. But that's actually not relevant for the rate-limitation. :facepalm: The only thing we care about is resetting the counter that limits the amount of data being produced. Therefore, resetting it when a new checkpoint is triggered is reasonable.